clickhouse-admin/types/src/lib.rs

```diff
@@ -121,12 +120,16 @@ impl ServerSettings {
             self.datastore_path.clone(),
         );
 
-        // Writing to a temporary file and then renaming it will ensure we
-        // don't end up with a partially written file after a crash
-        let mut f = NamedUtf8TempFile::new()?;
+        match create_dir(self.config_dir.clone()) {
+            Ok(_) => (),
+            Err(e) if e.kind() == ErrorKind::AlreadyExists => (),
+            Err(e) => return Err(e.into()),
+        };
+        let mut f =
+            File::create(self.config_dir.join("replica-server-config.xml"))?;
         f.write_all(config.to_xml().as_bytes())?;
         f.flush()?;
-        rename(f.path(), self.config_dir.join("replica-server-config.xml"))?;
```

Comment:
Sadly, this rename produces an error when running via the clickhouse-admin API inside an SMF service:

```console
{
  "request_id": "a529336d-e113-429f-b90e-5bb80f7912fc",
  "error_code": "Internal",
  "message": "clickward XML generation failure: Cross-device link (os error 18)"
}
```

Reply:
We still want to create a temporary file and rename, otherwise we can end up with partially written configurations that do not get corrected without support and/or a config change. There are ways to work around this, but atomic rename is typically the simplest. And it will help once we have generation numbers and need to write those to disk as well. 

My guess is that this is failing because `NamedUtf8TempFile` is putting the temp file on a `/tmpfs` device and you can't rename it to the zfs filesystem on an actual U.2 device. You should be able to fix this by not using `NamedUtf8TempFile` and instead just creating a file named `replica-server-config.xml.tmp` in `self.config_dir()` and then doing the atomic rename.

Reply:
To emphasize what can go wrong. If we partially write the file, yes we will get an error, and yes we can retry. But the config will automatically get loaded by clickhouse and may crash it or cause other issues. It's better to have a stable old config and only load valid configs.

Reply:
There's also the **atomicwrites** crate which is pretty good; e.g., https://github.com/oxidecomputer/omicron/blob/3d3f6d735e44c8f84b7b23d7e6ca3b213ac15355/dev-tools/openapi-manager/src/spec.rs#L482

Reply:
> There's also the **atomicwrites** crate which is pretty good; e.g.,
> 
> https://github.com/oxidecomputer/omicron/blob/3d3f6d735e44c8f84b7b23d7e6ca3b213ac15355/dev-tools/openapi-manager/src/spec.rs#L482

Oh nice. TIL

Reply:
Ohhh that's a very useful crate! I've updated the code